// solana-agent-kit/dist/actions/pythFetchPrice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const tools_1 = require("../tools");
const pythFetchPriceAction = {
    name: "PYTH_FETCH_PRICE",
    similes: [
        "get pyth price",
        "check pyth price",
        "pyth oracle price",
        "fetch from pyth",
        "pyth price feed",
        "oracle price",
    ],
    description: "Fetch the current price from a Pyth oracle price feed",
    examples: [
        [
            {
                input: {
                    priceFeedId: "Gnt27xtC473ZT2Mw5u8wZ68Z3gULkSTb5DuxJy7eJotD", // SOL/USD price feed
                },
                output: {
                    status: "success",
                    price: "23.45",
                    message: "Current price: $23.45",
                },
                explanation: "Get the current SOL/USD price from Pyth oracle",
            },
        ],
    ],
    schema: zod_1.z.object({
        priceFeedId: zod_1.z
            .string()
            .min(1)
            .describe("The Pyth price feed ID to fetch the price from"),
    }),
    handler: async (_agent, input) => {
        try {
            const priceFeedId = input.tokenId;
            const priceStr = await (0, tools_1.pythFetchPrice)(priceFeedId);
            return {
                status: "success",
                price: priceStr,
                message: `Current price: $${priceStr}`,
            };
        }
        catch (error) {
            return {
                status: "error",
                message: `Failed to fetch price from Pyth: ${error.message}`,
            };
        }
    },
};
exports.default = pythFetchPriceAction;
