// solana-agent-kit/dist/action/SolanaTokenDataByTickerAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaTokenDataByTickerAction = void 0;
const zod_1 = require("zod");
exports.SolanaTokenDataByTickerAction = {
    name: "solana_token_data_by_ticker",
    similes: ["get_token_data_by_ticker", "fetch_token_data_by_ticker"],
    description: `Get the token data for a given token ticker.

  Inputs:
  ticker: string, eg "USDC" (required)`,
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: {
                            text: JSON.stringify({
                                ticker: "SOL"
                            }),
                        },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: {
                    success: true,
                    data: {
                        name: "Solana",
                        symbol: "SOL",
                        decimals: 9,
                        supply: "1000000000"
                    },
                },
            },
        ],
    ],
    handler: async (agent, input) => {
        const tokenData = await agent.getTokenDataByTicker(input.ticker.trim());
        return {
            success: true,
            data: {
                tokenData
            }
        };
    },
    validate: async (context, ...args) => {
        const input = args[0];
        try {
            const schema = zod_1.z.object({
                ticker: zod_1.z.string(),
            });
            return schema.safeParse(input).success;
        }
        catch {
            return false;
        }
    },
};
