// solana-agent-kit/dist/action/SolanaGetMainDomainAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaGetMainDomainAction = void 0;
const web3_js_1 = require("@solana/web3.js");
const zod_1 = require("zod");
exports.SolanaGetMainDomainAction = {
    name: "solana_get_main_domain",
    similes: ["get_main_domain", "fetch_main_domain"],
    description: `Get the main/favorite domain for a given wallet address.

  Inputs:
  owner: string, eg "4Be9CvxqHW6BYiRAxW9Q3xu1ycTMWaL5z8NX4HR3ha7t" (required)`,
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: {
                            text: JSON.stringify({
                                owner: "4Be9CvxqHW6BYiRAxW9Q3xu1ycTMWaL5z8NX4HR3ha7t"
                            }),
                        },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: {
                    success: true,
                    data: { mainDomain: "example.sol" }
                },
            },
        ],
    ],
    handler: async (agent, input) => {
        const ownerPubkey = new web3_js_1.PublicKey(input.owner.trim());
        const mainDomain = await agent.getMainAllDomainsDomain(ownerPubkey);
        return {
            success: true,
            data: {
                domain: mainDomain,
                owner: input.owner
            }
        };
    },
    validate: async (context, ...args) => {
        const input = args[0];
        try {
            const schema = zod_1.z.object({
                owner: zod_1.z.string()
            });
            return schema.safeParse(input).success;
        }
        catch {
            return false;
        }
    },
};
