// solana-agent-kit/dist/action/SolanaRaydiumCreateClmmAction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaRaydiumCreateClmmAction = void 0;
const web3_js_1 = require("@solana/web3.js");
const decimal_js_1 = __importDefault(require("decimal.js"));
const anchor_1 = require("@coral-xyz/anchor");
const zod_1 = require("zod");
exports.SolanaRaydiumCreateClmmAction = {
    name: "raydium_create_clmm",
    similes: ["create_clmm", "raydium_create_clmm"],
    description: `Create a Concentrated Liquidity Market Maker (CLMM) pool.

  Inputs:
  mint1: string (required)
  mint2: string (required)
  configId: string (required) - stores pool info, id, index, protocolFeeRate, tradeFeeRate, tickSpacing, fundFeeRate
  initialPrice: number (required)
  startTime: number(seconds) (required)`,
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: {
                            text: JSON.stringify({
                                mint1: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
                                mint2: "So11111111111111111111111111111111111111112",
                                configId: "config_pubkey_here",
                                initialPrice: 1.5,
                                startTime: Math.floor(Date.now() / 1000)
                            }),
                        },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: {
                    success: true,
                    data: { txId: "transaction_hash_here" }
                },
            },
        ],
    ],
    handler: async (agent, input) => {
        const tx = await agent.raydiumCreateClmm(new web3_js_1.PublicKey(input.mint1), new web3_js_1.PublicKey(input.mint2), new web3_js_1.PublicKey(input.configId), new decimal_js_1.default(input.initialPrice), new anchor_1.BN(input.startTime));
        return {
            success: true,
            data: {
                transaction: tx,
                message: "Created Raydium CLMM pool successfully"
            }
        };
    },
    validate: async (context, ...args) => {
        const input = args[0];
        try {
            const schema = zod_1.z.object({
                mint1: zod_1.z.string(),
                mint2: zod_1.z.string(),
                configId: zod_1.z.string(),
                initialPrice: zod_1.z.number().positive(),
                startTime: zod_1.z.number().int()
            });
            return schema.safeParse(input).success;
        }
        catch {
            return false;
        }
    },
};
