// solana-agent-kit/dist/action/SolanaLendAssetAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaLendAssetAction = void 0;
const zod_1 = require("zod");
exports.SolanaLendAssetAction = {
    name: "solana_lend_asset",
    similes: ["lend_asset", "provide_liquidity"],
    description: `Lend idle USDC for yield using Lulo. (only USDC is supported)

  Inputs:
  amount: number, eg 1, 0.01 (required)`,
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: {
                            text: JSON.stringify({
                                amount: 100,
                            }),
                        },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: {
                    success: true,
                    data: { transactionId: "5G9f8..." },
                },
            },
        ],
    ],
    handler: async (agent, input) => {
        const tx = await agent.lendAssets(input.amount);
        return {
            success: true,
            data: {
                transaction: tx,
                amount: input.amount
            }
        };
    },
    validate: async (context, ...args) => {
        const input = args[0];
        try {
            const schema = zod_1.z.object({
                amount: zod_1.z.number().positive(),
            });
            return schema.safeParse(input).success;
        }
        catch {
            return false;
        }
    },
};
