// solana-agent-kit/dist/action/SolanaCreateImageAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaCreateImageAction = void 0;
const zod_1 = require("zod");
const create_image_1 = require("../tools/create_image");
exports.SolanaCreateImageAction = {
    name: "solana_create_image",
    similes: ["generate_image", "create_image"],
    description: "Create an image using OpenAI's DALL-E. Input should be a string prompt for the image.",
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: {
                            text: JSON.stringify({
                                prompt: "A futuristic cityscape",
                            }),
                        },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: {
                    success: true,
                    data: { imageUrl: "https://example.com/image.png" },
                },
            },
        ],
    ],
    handler: async (agent, input) => {
        const result = await (0, create_image_1.create_image)(agent, input.prompt);
        return {
            success: true,
            data: result,
        };
    },
    validate: async (context, ...args) => {
        const input = args[0];
        try {
            const schema = zod_1.z.object({
                prompt: zod_1.z.string(),
            });
            return schema.safeParse(input).success;
        }
        catch {
            return false;
        }
    },
};
