// solana-agent-kit/dist/action/SolanaBalanceAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaBalanceAction = void 0;
const web3_js_1 = require("@solana/web3.js");
const zod_1 = require("zod");
exports.SolanaBalanceAction = {
    name: "solana_balance",
    similes: ["check_balance", "get_wallet_balance"],
    description: `Get the balance of a Solana wallet or token account,

  If you want to get the balance of your wallet, you don't need to provide the tokenAddress.
  If no tokenAddress is provided, the balance will be in SOL.

  Inputs:
  tokenAddress: string, eg "So11111111111111111111111111111111111111112" (optional)`,
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: { text: "Get my balance" },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: { success: true, data: { balance: "100 SOL" } },
            },
        ],
    ],
    handler: async (agent, input) => {
        const convert = async () => {
            const publicKey = new web3_js_1.PublicKey(input.tokenAddress);
            const promise_string = await agent.getBalance(publicKey);
            const result = {
                success: true,
                data: { balance: promise_string },
            };
            return result;
        };
        return convert();
    },
    validate: async (context, ...args) => {
        const input = args[0];
        try {
            const schema = zod_1.z.string().optional();
            return schema.safeParse(input).success;
        }
        catch {
            return false;
        }
    },
};
