// solana-agent-kit/dist/action/SolanaGetAllTldsAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaGetAllTldsAction = void 0;
exports.SolanaGetAllTldsAction = {
    name: "solana_get_all_tlds",
    similes: ["get_all_tlds", "fetch_all_tlds"],
    description: "Get all active top-level domains (TLDs) in the AllDomains Name Service",
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: { text: "" },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: {
                    success: true,
                    data: { tlds: ["sol", "bonk", "blink"] }
                },
            },
        ],
    ],
    handler: async (agent) => {
        const tlds = await agent.getAllDomainsTLDs();
        return {
            success: true,
            data: {
                tlds
            }
        };
    },
    validate: async () => true, // No input to validate
};
