// solana-agent-kit/dist/action/SolanaRaydiumCreateAmmV4Action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaRaydiumCreateAmmV4Action = void 0;
const web3_js_1 = require("@solana/web3.js");
const anchor_1 = require("@coral-xyz/anchor");
const zod_1 = require("zod");
exports.SolanaRaydiumCreateAmmV4Action = {
    name: "raydium_create_ammV4",
    similes: ["create_amm_v4", "raydium_create_amm"],
    description: `Raydium's Legacy AMM that requires an OpenBook marketID.

  Inputs:
  marketId: string (required)
  baseAmount: number(int) (required)
  quoteAmount: number(int) (required)
  startTime: number(seconds) (required)`,
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: {
                            text: JSON.stringify({
                                marketId: "9vMJfxuKxXBoEa7rM12mYLMwTacLMLDJqHozw96WQL8i",
                                baseAmount: 1000000,
                                quoteAmount: 1000000,
                                startTime: Math.floor(Date.now() / 1000)
                            }),
                        },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: {
                    success: true,
                    data: { txId: "transaction_hash_here" }
                },
            },
        ],
    ],
    handler: async (agent, input) => {
        const tx = await agent.raydiumCreateAmmV4(new web3_js_1.PublicKey(input.marketId), new anchor_1.BN(input.baseAmount), new anchor_1.BN(input.quoteAmount), new anchor_1.BN(input.startTime));
        return {
            success: true,
            data: {
                transaction: tx,
                message: "Created Raydium AMM V4 pool successfully"
            }
        };
    },
    validate: async (context, ...args) => {
        const input = args[0];
        try {
            const schema = zod_1.z.object({
                marketId: zod_1.z.string(),
                baseAmount: zod_1.z.number().int().positive(),
                quoteAmount: zod_1.z.number().int().positive(),
                startTime: zod_1.z.number().int()
            });
            return schema.safeParse(input).success;
        }
        catch {
            return false;
        }
    },
};
