// solana-agent-kit/dist/action/SolanaRequestFundsAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaRequestFundsAction = void 0;
const zod_1 = require("zod");
exports.SolanaRequestFundsAction = {
    name: "solana_request_funds",
    similes: ["request_airdrop", "get_test_tokens"],
    description: "Request SOL from Solana faucet (devnet/testnet only)",
    examples: [
        [
            {
                input: {
                    message: {
                        id: "1",
                        content: {
                            text: JSON.stringify({
                                amount: 1,
                            }),
                        },
                        userId: "user1",
                        timestamp: Date.now(),
                    },
                },
                output: {
                    success: true,
                    data: { transactionId: "5G9f8..." },
                },
            },
        ],
    ],
    handler: async (agent, input) => {
        await agent.requestFaucetFunds();
        return {
            success: true,
            data: {
                message: "Successfully requested faucet funds",
                network: agent.connection.rpcEndpoint.split("/")[2]
            }
        };
    },
    validate: async (context, ...args) => {
        const input = args[0];
        try {
            const schema = zod_1.z.object({
                amount: zod_1.z.number(),
            });
            return schema.safeParse(input).success;
        }
        catch {
            return false;
        }
    },
};
