// @cks-systems/manifest-sdk/dist/esm/src/ui_wrapper/instructions/ClaimSeatUnused.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
/**
 * @category Instructions
 * @category ClaimSeatUnused
 * @category generated
 */
export const ClaimSeatUnusedStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.u8]], 'ClaimSeatUnusedInstructionArgs');
export const claimSeatUnusedInstructionDiscriminator = 1;
/**
 * Creates a _ClaimSeatUnused_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category ClaimSeatUnused
 * @category generated
 */
export function createClaimSeatUnusedInstruction(accounts, programId = new web3.PublicKey('UMnFStVeG1ecZFc2gc5K3vFy3sMpotq8C91mXBQDGwh')) {
    const [data] = ClaimSeatUnusedStruct.serialize({
        instructionDiscriminator: claimSeatUnusedInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: accounts.manifestProgram,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.owner,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: accounts.market,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.payer,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: accounts.wrapperState,
            isWritable: true,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
