// @cks-systems/manifest-sdk/dist/esm/src/ui_wrapper/instructions/EditOrder.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as splToken from '@solana/spl-token';
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
/**
 * @category Instructions
 * @category EditOrder
 * @category generated
 */
export const EditOrderStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.u8]], 'EditOrderInstructionArgs');
export const editOrderInstructionDiscriminator = 3;
/**
 * Creates a _EditOrder_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category EditOrder
 * @category generated
 */
export function createEditOrderInstruction(accounts, programId = new web3.PublicKey('UMnFStVeG1ecZFc2gc5K3vFy3sMpotq8C91mXBQDGwh')) {
    const [data] = EditOrderStruct.serialize({
        instructionDiscriminator: editOrderInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: accounts.wrapperState,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.owner,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.traderTokenAccount,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.market,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.vault,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.mint,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.tokenProgram ?? splToken.TOKEN_PROGRAM_ID,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.manifestProgram,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.payer,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: accounts.baseMint,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.baseGlobal,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.baseGlobalVault,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.baseMarketVault,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.baseTokenProgram,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteMint,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteGlobal,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteGlobalVault,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteMarketVault,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteTokenProgram,
            isWritable: true,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
