// solana-agent-kit/dist/tools/manifest_create_market.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.manifestCreateMarket = manifestCreateMarket;
const manifest_sdk_1 = require("@cks-systems/manifest-sdk");
const web3_js_1 = require("@solana/web3.js");
async function manifestCreateMarket(agent, baseMint, quoteMint) {
    const marketKeypair = web3_js_1.Keypair.generate();
    const FIXED_MANIFEST_HEADER_SIZE = 256;
    const createAccountIx = web3_js_1.SystemProgram.createAccount({
        fromPubkey: agent.wallet.publicKey,
        newAccountPubkey: marketKeypair.publicKey,
        space: FIXED_MANIFEST_HEADER_SIZE,
        lamports: await agent.connection.getMinimumBalanceForRentExemption(FIXED_MANIFEST_HEADER_SIZE),
        programId: new web3_js_1.PublicKey("MNFSTqtC93rEfYHB6hF82sKdZpUDFWkViLByLd1k1Ms"),
    });
    const createMarketIx = manifest_sdk_1.ManifestClient["createMarketIx"](agent.wallet.publicKey, baseMint, quoteMint, marketKeypair.publicKey);
    const tx = new web3_js_1.Transaction();
    tx.add(createAccountIx);
    tx.add(createMarketIx);
    const signature = await (0, web3_js_1.sendAndConfirmTransaction)(agent.connection, tx, [
        agent.wallet,
        marketKeypair,
    ]);
    return [signature, marketKeypair.publicKey.toBase58()];
}
